HashMapStylePropertyMap is readonly and shouldn't be subclassing StylePropertyMap

cdumez opened the pull request:
#### b902c3b0e0c7e62d8bdf627e5acee6b9cf7acaa4
<pre>
HashMapStylePropertyMap is readonly and shouldn&apos;t be subclassing StylePropertyMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=247012">https://bugs.webkit.org/show_bug.cgi?id=247012</a>

Reviewed by Youenn Fablet.

HashMapStylePropertyMap is readonly and shouldn&apos;t be subclassing StylePropertyMap.
Instead, introduce a MainThreadStylePropertyMapReadonly and have both
HashMapStylePropertyMap and StylePropertyMap subclass it.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.cpp: Copied from Source/WebCore/css/typedom/StylePropertyMap.cpp.
(WebCore::MainThreadStylePropertyMapReadOnly::documentFromContext):
(WebCore::MainThreadStylePropertyMapReadOnly::get const):
(WebCore::MainThreadStylePropertyMapReadOnly::getAll const):
(WebCore::MainThreadStylePropertyMapReadOnly::has const):
(WebCore::MainThreadStylePropertyMapReadOnly::shorthandPropertyValue const):
* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.h: Copied from Source/WebCore/css/typedom/StylePropertyMap.h.
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::documentFromContext): Deleted.
(WebCore::StylePropertyMap::shorthandPropertyValue const): Deleted.
(WebCore::StylePropertyMap::get const): Deleted.
(WebCore::StylePropertyMap::getAll const): Deleted.
(WebCore::StylePropertyMap::has const): Deleted.
* Source/WebCore/css/typedom/StylePropertyMap.h:
* Source/WebCore/html/CustomPaintImage.cpp:

Canonical link: <a href="https://commits.webkit.org/256012@main">https://commits.webkit.org/256012@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b3b9a221e4954189cae71249164c62527ce9c0e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94218 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3410 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/25047 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103900 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164176 "Found 1 new test failure: http/wpt/webcodecs/videoFrame-serialization.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98219 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3437 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31650 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86576 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99921 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99874 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2465 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80630 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29508 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84384 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72430 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38006 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17879 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35890 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19151 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39768 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41734 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1959 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41711 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38380 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.h:
```diff
@@ -0,0 +1,52 @@
+public:
+    ExceptionOr<RefPtr<CSSStyleValue>> get(ScriptExecutionContext&, const AtomString&) const final;
+    ExceptionOr<Vector<RefPtr<CSSStyleValue>>> getAll(ScriptExecutionContext&, const AtomString&) const final;
+    ExceptionOr<bool> has(ScriptExecutionContext&, const AtomString&) const final;
```
Do they need to be public?

cdumez replied:
While they currently don't need to be, those are technically public API and it doesn't hurt to expose them.
We have this bad pattern (IMO) in WebKit to override public functions and mark the overrides as private.
The reason I think it is bad is because I then see people introducing new code which casts objects to their base class, just so that they can call these virtual functions. Not only does this casting look ugly but it also likely prevents the compiler from de-virtualizing the function call.

cdumez committed:
HashMapStylePropertyMap is readonly and shouldn't be subclassing StylePropertyMap
https://bugs.webkit.org/show_bug.cgi?id=247012

Reviewed by Youenn Fablet.

HashMapStylePropertyMap is readonly and shouldn't be subclassing StylePropertyMap.
Instead, introduce a MainThreadStylePropertyMapReadonly and have both
HashMapStylePropertyMap and StylePropertyMap subclass it.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.cpp: Copied from Source/WebCore/css/typedom/StylePropertyMap.cpp.
(WebCore::MainThreadStylePropertyMapReadOnly::documentFromContext):
(WebCore::MainThreadStylePropertyMapReadOnly::get const):
(WebCore::MainThreadStylePropertyMapReadOnly::getAll const):
(WebCore::MainThreadStylePropertyMapReadOnly::has const):
(WebCore::MainThreadStylePropertyMapReadOnly::shorthandPropertyValue const):
* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.h: Copied from Source/WebCore/css/typedom/StylePropertyMap.h.
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::documentFromContext): Deleted.
(WebCore::StylePropertyMap::shorthandPropertyValue const): Deleted.
(WebCore::StylePropertyMap::get const): Deleted.
(WebCore::StylePropertyMap::getAll const): Deleted.
(WebCore::StylePropertyMap::has const): Deleted.
* Source/WebCore/css/typedom/StylePropertyMap.h:
* Source/WebCore/html/CustomPaintImage.cpp:

Canonical link: https://commits.webkit.org/256012@main
